src/waveform/renderers/allshader/waveformrenderertextured.cpp

```diff
@@ -0,0 +1,515 @@
+    int dataSize = 0;
+    const WaveformData* data = nullptr;
+
+    ConstWaveformPointer pWaveform = m_waveformRenderer->getWaveform();
+    if (pWaveform) {
+        dataSize = pWaveform->getDataSize();
+        if (dataSize > 1) {
+            data = pWaveform->data();
+        }
+    }
+
+    glEnable(GL_TEXTURE_2D);
+
+    if (m_textureId == 0) {
+        glGenTextures(1, &m_textureId);
+
+        int error = glGetError();
+        if (error) {
+            qDebug() << "WaveformRendererTextured::loadTexture - m_textureId"
+                     << m_textureId << "error" << error;
+        }
+    }
+
+    glBindTexture(GL_TEXTURE_2D, m_textureId);
+
+    int error = glGetError();
+    if (error) {
+        qDebug() << "WaveformRendererTextured::loadTexture - bind error" << error;
+    }
+
+    glTexParameterf(GL_TEXTURE_2D, GL_TEXTURE_MAG_FILTER, GL_NEAREST);
+    glTexParameterf(GL_TEXTURE_2D, GL_TEXTURE_MIN_FILTER, GL_NEAREST);
+
+    if (pWaveform != nullptr && data != nullptr) {
+        // Waveform ensures that getTextureSize is a multiple of
+        // getTextureStride so there is no rounding here.
+        int textureWidth = pWaveform->getTextureStride();
+        int textureHeight = pWaveform->getTextureSize() / pWaveform->getTextureStride();
+
+        glTexImage2D(GL_TEXTURE_2D,
+                0,
+                GL_RGBA,
+                textureWidth,
+                textureHeight,
+                0,
+                GL_RGBA,
+                GL_UNSIGNED_BYTE,
+                data);
+        int error = glGetError();
+        if (error) {
+            qDebug() << "WaveformRendererTextured::loadTexture - glTexImage2D error" << error;
+        }
+    } else {
+        glDeleteTextures(1, &m_textureId);
+        m_textureId = 0;
+    }
+
+    glDisable(GL_TEXTURE_2D);
+
+    return true;
+}
+
+void WaveformRendererTextured::createGeometry() {
+    if (m_unitQuadListId != -1) {
+        return;
+    }
+
+    glMatrixMode(GL_PROJECTION);
+    glLoadIdentity();
+    glOrtho(-1.0, 1.0, -1.0, 1.0, -10.0, 10.0);
+    glMatrixMode(GL_MODELVIEW);
+    glLoadIdentity();
+
+    m_unitQuadListId = glGenLists(1);
+    glNewList(m_unitQuadListId, GL_COMPILE);
+    {
+        glBegin(GL_QUADS);
+        {
+            glTexCoord2f(0.0, 0.0);
+            glVertex3f(-1.0f, -1.0f, 0.0f);
+
+            glTexCoord2f(1.0, 0.0);
+            glVertex3f(1.0f, -1.0f, 0.0f);
+
+            glTexCoord2f(1.0, 1.0);
+            glVertex3f(1.0f, 1.0f, 0.0f);
+
+            glTexCoord2f(0.0, 1.0);
+            glVertex3f(-1.0f, 1.0f, 0.0f);
+        }
+        glEnd();
+    }
+    glEndList();
+}
+
+void WaveformRendererTextured::createFrameBuffers() {
+    const float devicePixelRatio = m_waveformRenderer->getDevicePixelRatio();
+    // We create a frame buffer that is 4x the size of the renderer itself to
+    // "oversample" the texture relative to the surface we're drawing on.
+    constexpr int oversamplingFactor = 4;
```

Comment:
how is that different from multisampling?

Reply:
I don't know, I just took the legacy GLSL renderer code as-is and adapted the minimum so it could be used together with the rest of the allshader layers.